File: native/core-swift/Sources/C15tCore/VendorListNarrowing.swift
// A publisher's vendor scope, applied to a list that arrived whole.
//
// Two web functions do this job, and both are the oracle here: `narrowGVLToVendors` in
// `packages/iab/src/tcf/fetch-gvl.ts`, and `narrowToVendorIds` in
// `packages/backend/src/http/gvl.ts`. Both exist because the scope cannot always travel
// with the document. The GVL endpoint applies `vendorIds` as a filter, which is what keeps
// a scoped publisher off the full list, but the filter only travels so far: the backend
// caps what it will put in a request line at `MAX_GVL_QUERY_VENDOR_IDS` and fetches the
// list whole above that, `createIAB` in `packages/iab` asks the endpoint for its scope and
// prunes whatever comes back in case the endpoint ignored the parameter, and neither path
// lets an unpruned document reach the surface a subject reads. Serving every GVL vendor to a
// visitor who was only ever shown a publisher's own partners is a disclosure failure, not
// a cosmetic one, so the prune is the part that carries the promise and the filter is only
// the bytes-saving optimisation in front of it.
//
// This file narrows a decoded list rather than the served bytes, for the same reason the
// web narrows a parsed object: by the time the core has a vendor list it holds a
// ``GlobalVendorList``, and `StoredEnvelope` writes what that value contains. One prune on
// the value therefore reaches the snapshot, the bridge JSON, and the envelope built from
// both. A second prune on raw JSON would not be a safety net on top of that, it would be a
// second answer -- and `StoredEnvelope` refuses an envelope whose stored key paths do not
// come back out of the typed value, so raw bytes pruned out of step with the value they
// were read from is the one edit that turns a stored list into nothing stored.

extension GlobalVendorList {
    /// The list with ``vendors`` kept down to `vendorIds`, and every other field untouched.
    ///
    /// The scope is a publisher's declaration about which vendors this deployment may
    /// disclose, and a list that arrived carrying more than that has to be reduced before
    /// anything renders from it or encodes from it. Purposes, features, special features,
    /// stacks, data categories and the version fields all come back exactly as served:
    /// they describe the framework, not the publisher's partner list, and the dialog needs
    /// their names while ``TcVendorList`` needs the two versions. Pruning a purpose alongside
    /// a vendor is how a device ends up rendering a row nobody served and advertising a
    /// vendor-list version no one agreed to.
    ///
    /// Two rules hold the prune honest, both taken from the oracles:
    ///
    /// - A scope of `nil` or `[]` is no scope. Both web functions return the list
    ///   untouched in that case, so a deployment that declared no scope keeps a complete
    ///   list instead of coming back with an empty drawer that a subject has to be told is
    ///   not a refusal.
    /// - Entries are chosen by dictionary key, never by reading an `id` out of an entry
    ///   body. ``GlobalVendorList/read(from:)`` already keys a dense-array publication by
    ///   each entry's own `id`, so the key is the id the document was read as; letting a
    ///   body `id` take part would let an entry whose two disagree answer for a vendor the
    ///   publisher did not name.
    ///
    /// Nothing is ever added: a scope id with no entry in the served list simply does not
    /// appear in the result. That is the point of pruning a served document rather than
    /// rebuilding one from the scope, and it is why an out-of-range or publisher-custom id
    /// costs nothing here -- a vendor the framework never assigned has no entry to survive,
    /// and inventing one would put a name on a drawer that no GVL revision stands behind.
    ///
    /// - Parameter vendorIds: The declared scope. `nil` or empty keeps every vendor served.
    /// - Returns: A list whose `vendors` are the served entries inside `vendorIds`, keyed and
    ///   valued exactly as served, with every other field as served.
    func narrowed(toVendorIds vendorIds: [Int]?) -> GlobalVendorList {
        guard let vendorIds, !vendorIds.isEmpty else { return self }
        let allowed = Set(vendorIds)
        // Filtering the served map, in both directions: only keys the scope names survive,
        // and the values that survive are carried over whole rather than rebuilt.
        let keptVendors = vendors.filter { allowed.contains($0.key) }
        // Field by field because every field on this type is a `let`. An entry here is how
        // a field added to ``GlobalVendorList`` reaches a narrowed list; leaving one out
        // quietly hands the dialog a default instead of what the backend served.
        return GlobalVendorList(
            gvlSpecificationVersion: gvlSpecificationVersion,
            vendorListVersion: vendorListVersion,
            tcfPolicyVersion: tcfPolicyVersion,
            lastUpdated: lastUpdated,
            purposes: purposes,
            specialPurposes: specialPurposes,
            features: features,
            specialFeatures: specialFeatures,
            stacks: stacks,
            dataCategories: dataCategories,
            vendors: keptVendors
        )
    }
}

File: native/core-android/c15t-core/src/main/kotlin/com/c15t/core/tc/VendorListNarrowing.kt
package com.c15t.core.tc

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject

/**
 * The publisher's vendor scope, applied to a list that arrived whole.
 *
 * c15t sends the scope upstream rather than filtering after the fact: `gvlRequestUrl` in
 * `packages/backend/src/http/gvl.ts` puts `vendorIds` on the GVL request, `narrowToVendorIds` in that
 * file then runs the filter over the document that came back before it is cached, and
 * `narrowGVLToVendors` in `packages/iab/src/tcf/fetch-gvl.ts` runs it over a list that reached the
 * browser by another route so server-resolved state and a client fetch disclose the same partners.
 * The filter has to run somewhere, and a device holding the complete list is one careless render away
 * from disclosing every GVL vendor to a subject who was only ever shown a publisher's own partners.
 *
 * The prune stops at `vendors`, which is what both web functions do and is not an oversight about the
 * rest of the document. Purposes, special purposes, features, special features, stacks, data
 * categories and the two version numbers describe the framework rather than the audience: a list
 * narrowed to three vendors still has to be able to say what purpose 7 means and which
 * `tcfPolicyVersion` its TC String is graded against. Thinning the unreferenced ones for tidiness
 * would also install a second, silent reading rule -- `processPurposes` in
 * `packages/iab/src/headless/dialog-data.ts` walks the served `purposes` record to draw the purposes
 * leg of a dialog, and a copy of it pruned here would change what that dialog claims independently of
 * the scope anybody configured.
 *
 * An empty scope means "no scope configured", not "show nobody". That is the web's reading twice
 * over: `narrowGVLToVendors` hands the list back for `vendorIds.length === 0`, and `gvlRequestUrl`
 * does not append the parameter for an empty list at all, so the document on the wire is the whole
 * one. A host that means no vendors has to say so somewhere other than here. This is the fail-open
 * half of the pair, which is why [narrowVendorListElement] and this function read the empty case the
 * same way and say so out loud rather than leaving it to a caller's memory.
 */
fun GlobalVendorList.narrowToVendorIds(vendorIds: Collection<Int>?): GlobalVendorList {
	if (vendorIds.isNullOrEmpty()) {
		return this
	}
	val allowed = vendorIds.toSet()
	return copy(vendors = vendors.filterKeys { id -> id in allowed })
}

/**
 * [GlobalVendorList.narrowToVendorIds] for bytes that have not been decoded: a served vendor-list
 * document, in the shape [GlobalVendorListJson] reads it or a bridge forwards it.
 *
 * The element form exists because narrowing is not always a decision a caller gets to make after a
 * decode. `GlobalVendorListJson.read` refuses a document this build cannot name whole, and a body
 * forwarded on to JavaScript never goes through a decode at all, so a caller that holds the raw `gvl`
 * value and wants one scope applied needs the answer at this level. The backend stands in the same
 * spot: it narrows `parsed.output` before `cache.set`, precisely so a wide list cannot survive in the
 * cache under a key that promises a scoped one.
 *
 * Nothing is pruned, and the input comes back as the same instance, when the scope is null or empty --
 * for the reason above -- and when there is no `vendors` record to prune: [element] null, a JSON
 * `null`, a string, an array, an object with no `vendors` key, or one whose `vendors` is not an
 * object. Identity is the point rather than an optimisation: it keeps "did this call rewrite my
 * bytes?" off every caller's list of questions.
 *
 * Whether a document is worth holding is not this function's decision, so no part of the acceptance
 * rules is re-checked here. [GlobalVendorListJson.read] owns that, and re-deriving it would make the
 * two paths disagree: a document refused downstream is null either way, while a document skipped here
 * because it failed a test it was never handed leaves the tier of bytes a caller forwarded unscoped.
 *
 * The surviving keys are the ones the served keys name, compared as written. A GVL's numbered
 * records are objects keyed by a number written as a string, and both web filters build the allowlist
 * with `vendorIds.map(String)` and test it against the keys `Object.entries` hands them; this compares
 * the same two things, so a key served as `"042"`, or with a space in front of it, is not a vendor the
 * scope names and is pruned away. An entry's own `id` field never gets a vote, which is the key's
 * authority recorded at the top of [GlobalVendorList] and again on `toTcVendor` in
 * `TcVendorListFromGvl.kt`: an entry that reads a different number in its body than the key it lives
 * under cannot use a scope as the occasion to rename itself into somebody else's disclosure.
 *
 * An id the served document does not carry contributes nothing, and no entry is written to satisfy
 * the scope -- inventing a vendor out of a number a publisher configured is contract rule 5 with a
 * friendlier face, and a fabricated entry would reach [toTcVendorList] as declarations no company
 * ever published. Siblings keep their position and value untouched, `vendors` keeps the served order
 * of what survives, and the pruned record can legitimately come back empty, which is the same answer
 * the web's filter gives and reads as no vendors downstream. Untouched values are shared and not
 * copied, which costs nothing because a [JsonElement] is immutable.
 */
fun narrowVendorListElement(element: JsonElement?, vendorIds: Collection<Int>?): JsonElement? {
	if (vendorIds.isNullOrEmpty()) {
		return element
	}
	val document = element as? JsonObject ?: return element
	val vendors = document[VENDORS_KEY] as? JsonObject ?: return element
	val allowed = vendorIds.mapTo(HashSet(vendorIds.size)) { id -> id.toString() }
	val pruned = JsonObject(vendors.filterKeys { key -> key in allowed })
	return JsonObject(document + (VENDORS_KEY to pruned))
}

/** The one record this file narrows. Its siblings are served data and stay as they arrived. */
private const val VENDORS_KEY = "vendors"
